### ISSUE TITLE
DFD usage and implementation

### ISSUE THREAD

**Issue by lexaknyazev** [KhronosGroup/KTX-Specification#64]
_The following feedback is based on DFD revision 1.2._

The DFD spec gives a lot of flexibility for providing formal definitions of "unknown" pixel formats such as uneven or padded bit widths, Bayer patterns, etc. Many DFD options duplicate fixed `vkFormat` definitions, so this leads to lots of implementation corner cases. 

DFD fields are:
#### color_model
```
    KHR_DF_MODEL_UNSPECIFIED
    KHR_DF_MODEL_RGBSDA
    KHR_DF_MODEL_YUVSDA
    KHR_DF_MODEL_YIQSDA
    KHR_DF_MODEL_LABSDA
    KHR_DF_MODEL_CMYKA
    KHR_DF_MODEL_XYZW
    KHR_DF_MODEL_HSVA_ANG
    KHR_DF_MODEL_HSLA_ANG
    KHR_DF_MODEL_HSVA_HEX
    KHR_DF_MODEL_HSLA_HEX
    KHR_DF_MODEL_YCGCOA
    KHR_DF_MODEL_YCCBCCRC
    KHR_DF_MODEL_ICTCP
    KHR_DF_MODEL_CIEXYZ
    KHR_DF_MODEL_CIEXYY
    KHR_DF_MODEL_BC1A
    KHR_DF_MODEL_BC2
    KHR_DF_MODEL_BC3
    KHR_DF_MODEL_BC4
    KHR_DF_MODEL_BC5
    KHR_DF_MODEL_BC6H
    KHR_DF_MODEL_BC7
    KHR_DF_MODEL_ETC1
    KHR_DF_MODEL_ETC2
    KHR_DF_MODEL_ASTC
```
Obviously, last ten options have 1:1 mappings with the corresponding compressed formats. Uncompressed formats can somewhat freely choose `color_model` from 16 other options and the application behavior is out of scope of KTX conformance. I think it's safe to assume that overwhelming majority of writers would use `RGBSDA` while most readers would ignore this filed altogether.

#### color_primaries
```
    KHR_DF_PRIMARIES_UNSPECIFIED
    KHR_DF_PRIMARIES_BT709 = KHR_DF_PRIMARIES_SRGB
    KHR_DF_PRIMARIES_BT601_EBU
    KHR_DF_PRIMARIES_BT601_SMPTE
    KHR_DF_PRIMARIES_BT2020
    KHR_DF_PRIMARIES_CIEXYZ
    KHR_DF_PRIMARIES_ACES
    KHR_DF_PRIMARIES_ACESCC
    KHR_DF_PRIMARIES_NTSC1953
    KHR_DF_PRIMARIES_PAL525
    KHR_DF_PRIMARIES_DISPLAYP3
    KHR_DF_PRIMARIES_ADOBERGB
```
Default choice should be `KHR_DF_PRIMARIES_SRGB`.

#### transfer_function
```
    KHR_DF_TRANSFER_UNSPECIFIED
    KHR_DF_TRANSFER_LINEAR
    KHR_DF_TRANSFER_SRGB
    KHR_DF_TRANSFER_ITU
    KHR_DF_TRANSFER_NTSC
    KHR_DF_TRANSFER_SLOG
    KHR_DF_TRANSFER_SLOG2
    KHR_DF_TRANSFER_BT1886
    KHR_DF_TRANSFER_HLG_OETF
    KHR_DF_TRANSFER_HLG_EOTF
    KHR_DF_TRANSFER_PQ_EOTF
    KHR_DF_TRANSFER_PQ_OETF
    KHR_DF_TRANSFER_DCIP3
    KHR_DF_TRANSFER_PAL_OETF
    KHR_DF_TRANSFER_PAL625_EOTF
    KHR_DF_TRANSFER_ST240
    KHR_DF_TRANSFER_ACESCC
    KHR_DF_TRANSFER_ACESCCT
    KHR_DF_TRANSFER_ADOBERGB
```
This one is quite difficult to implement properly. Hardware consistently supports only linear and sRGB filtering, with the latter available only for some specific formats (that have `_SRGB` in their names). Any other combination of DFD's `transfer_function` with `vkFormat` would require manual texture filtering which is out of scope for many platforms.

#### flags (alpha premultiplication)
Requires software pre-processing since this operation isn't available as an option on texture upload.

#### texel_block_dimension_0-3
These values are fixed for known `vkFormat`s.

#### bytes_plane_0-7
All Vulkan 1.0 formats has only 1 plane, so these fields are mostly useless. Moreover, these fields are not sufficient to describe the layouts of "multi-plane" formats from `VK_KHR_sampler_ycbcr_conversion` extension (part of Vulkan 1.1).

### Per-channel fields
_These are called "Sample Information" in the DFD spec._
#### bit_offset / bit_length
Channel data location. Fixed for all known `vkFormat`s.
#### channel_type
Contains mapping to color model (e.g. red or depth) and some typing flags (e.g. float). Fixed for all known `vkFormat`s when `color_model` is `RGBDSA`.
#### sample_position_0-3
Spatial location within texel block. Makes sense mostly for formats with downsampled channels
#### sample_lower/upper
Range remapping. Used for some video formats.

---

Conclusions:
1. Unless someone needs to signal an uncommon color model (e.g. CMYK), primaries, or transfer function, DFD doesn't provide any additional information for regular Vulkan 1.0 formats. And even in case of uncommon color model, there's no reason to reiterate each channel's information.
2. There are few cases when DFD provides a little extra information for Vulkan 1.0 compressed formats:
   - A `_FLOAT` flag with ASTC that signals the presence of HDR blocks and `texel_block_dimension_2` to describe 3D blocks. These would become less relevant once we have corresponding `vkFormat` enums.
   - An `ETC1` color model that signals that texture data doesn't contain ETC2 modes (ETC2 is a superset of ETC1, so there's no separate `vkFormat` enum).
   - Both these cases are covered with `KTXglFormat` with much less implementation complexity.
3. Only a subset of new multi-plane formats introduced in Vulkan 1.1 (initially from `VK_KHR_sampler_ycbcr_conversion`) can be described with DFD. At the same time, for many DFD options there are no API support.

---

In theory, DFD could be used as the only source of texture format. This would require readers to understand all DFD fields and being able to find the closest available API match or to perform some conversion. Such feature would considerably raise the complexity of the loaders. Writers would have to do the same in the other direction. This DFD / `vkFormat` redundancy hits the same points as `vkFormat` / `glFormat` discussion.

**Comment by MarkCallow** [comment]
I have 2 reasons for including the DFD:

1.  To provide the color space information. Color primaries are easily transformed in a shader. It's just another 4x4 transform. The transfer function is indeed an issue. Restricting to linear or sRGB for WebGL/glTF use makes complete sense. Some users want to do everything themselves though including handling the transfer function.
2.  To allow image processing outside of the graphics APIs to be performed on the image data without the processing tools having to be familiar with the details of all the VkFormats. Mostly this will happen during authoring or subsequent tuning of the images.

**Comment by lexaknyazev** [comment]
> without the processing tools having to be familiar with the details of all the VkFormats

Unless such tools output `VK_FORMAT_UNKNOWN`, they would need to choose what to write to vkFormat field, i.e., they would have to understand mappings between DFD and vkFormat.

**Comment by MarkCallow** [comment]
> Unless such tools output `VK_FORMAT_UNKNOWN`

My idea is that such tools will process the image data without changing its format. Applying filters and special effects and such.

**Comment by lexaknyazev** [comment]
A typical texture workflow goes like this. Initial data is usually authored in some high-end software like Photoshop or Substance using their native formats and saved as PNG or OpenEXR. Then, those textures are encoded to GPU-friendly formats (usually compressed). At this point, export tools would have to understand both DFD and `vkFormat` to produce KTX2 files. Filters and effects are usually applied to original data to preserve as much quality as possible.

I'm not aware of any tool that can directly edit arbitrary-defined texture data. For example, a legit DFD can describe red-green-alpha-depth pixel format where red and alpha are floats with shared exponent, while green and depth are normalized. Developing a tool that can correctly handle such layouts in all their glory would take much more efforts than baking-in a somewhat fixed set of familiar GPU formats.

---

DFD provides a subset of previously rejected "deep pixels" feature, so I'd ask a more broad question: does KTX2 aim to be GPU-ready texture container or a "universal" texture container?

To make the current design more reliable, I'd change DFD spec a bit (could it be done as DFD&nbsp;1.3?). Namely, I'd say that when `vkFormat` is not unknown, DFD should omit sample information. Maybe, a proper way forward would be to include `vkFormat` into DFD (2.0?) thus making its structure less redundant.

**Comment by fluppeteer** [comment]
Sorry, failed to spot this was still current. Writer coming for the DFD and some mappings, but the KDFS is completely independent of Vulkan (other than providing a documentation source for some information); I'm very keen that vkFormats don't get included. The point of the DFD is to avoid having a long arbitrary list of format enums that need mapping in order to understand them.

**Comment by lexaknyazev** [comment]
To use files with APIs, a "long arbitrary list of format enums" still has to exist somewhere. In our case, all KTX writers have to correctly write both: `vkFormat` and `DFD`. Also, it's not clear what readers should do with DFD's sample information when it doesn't match `vkFormat` (and to verify that, they have to contain all mappings as well).

**Comment by fluppeteer** [comment]
APIs that have chosen to define their formats as an arbitrary long list of format enums, yes. :-) My experience has been that both drivers and software decoders have to take the enums and apply multiple switch statements to break them down along the lines shown in the DFD. That's not to say that this is an inappropriate thing to do - the number of formats actually supported by most APIs is a small subset of what the DFD can represent, and is small enough to compress into enums.

KTX is designed to allow fast import and export (transport?) of textures from Khronos graphics APIs; it makes complete sense that it should have a vested interest in what the vkFormat should look like, and mappings involving this subset are forthcoming (honestly, lots of work still ongoing whenever I have free time). The KDFS treats that translation as peripheral.

**Comment by MarkCallow** [comment]
Happy to limit color model to RGBSDA, perhaps YUVSDA, and all the block-compressed formats.

color primaries should be free, as conversion is generally easy. Agree default should be BT709/SRGB.

Transfer function we should limit to SRGB or LINEAR give the difficulty of using others in the graphics APIs.

Isn't `alpha premultiplication` a signal to shaders as to how they should do alpha blending? We certainly don't want to encouraging unmultiply operations on loading. They are lossy operations.

Another piece of info the DFD provides is identifying the ETC1S subset of ETC1 as well as ETC1 from ETC2.

We might need the DFD for the wide transcodable format. It might not be equivalent to any Vulkan format.

**Comment by lexaknyazev** [comment]
> identifying the ETC1S subset of ETC1

Is it already in the spec?

**Comment by fluppeteer** [comment]
ETC1S is coming soon (Mark jumped the gun a bit); watch this space.

Alpha premultiplication indicates that the non-alpha channels have been pre-scaled by alpha. If you want to extract the colour that they'd contain were the surface opaque, then yes, you need to re-divide by alpha (which is troublesome), but it's slightly rare that you'd want to take a transparent surface and ignore its transparency. I'd hope that anyone using a premultiplied alpha surface is alpha-aware; if there were no support then we'd have to assume textures generated by typical blend operations were demultiplied prior to storage (unless they're fully opaque, in which case premultiplication is irrelevant).

**Comment by MarkCallow** [comment]
Please look at PR #64 and check I have the restrictions correct and haven't missed any you think should be there.

I don't see a reason to restrict alpha pre-multiplication so I haven't. If set, a shader would need to change how it does alpha blending but I don't see why it would require an operation on texture upload.

**Comment by fluppeteer** [comment]
Belated apologies that I've only just been looking back at this. Alexey, when you say:

> Only a subset of new multi-plane formats introduced in Vulkan 1.1 (initially from VK_KHR_sampler_ycbcr_conversion) can be described with DFD.

what did you mean? Additional information is required (in that things like where the planes are stored is outside the "format"), but I'm not aware of there being an inability to describe anything. If there is, I should fix it.

**Comment by lexaknyazev** [comment]
> Additional information is required

That's the issue. There's no place for such information in the KTX header yet. We thought (initially) that DFD will handle it.

**Comment by MarkCallow** [comment]
Fixed by PR #75.
